### ISSUE TITLE
Jeffrey Stuff to Work On

### ISSUE THREAD

**Issue by ghost** [GoTeamEpsilon/OpenEMR-AWS-Guide#27]
Hey Jeffrey:

- Per our Slack convo I created this issue just to track what you can work on ability-wise / what you'd like to work on.
- Maybe I'm projecting too much here, but the biggest issue starting out for me was I would see these vague issues and be like "wtf am I supposed to do with that".
- I would like these issues / tasks assigned to you to be very clear in how they can be achieved and shorter too...kind of like a quest you'd take in an RPG if you only had an hour left before bed.

1. Comment below on any IT skills you are good at / can do
2. Comment below on any IT skills you want to learn more about

@MatthewVita also after you respond Jeffrey, Matt we can take a look and maybe there is another OS project we can start in parallel that matches to Jeffrey's skills? Or maybe just work on Version 2.0 of this since @Jeffrey-P-McAteer I think you said you know Docker which I AM THE WORST AT.

**Comment by ghost** [comment]
3. Also what are your career goals?? Like what IT job do you want?? Haha I should have started with that that's a big one.

**Comment by Jeffrey-P-McAteer** [comment]
(I'm moving my slack response here)
<ctrl+p>
I'm halfway through college, but most of my education comes from my tinkering about with things. IT skills - I know quite a number of languages, from common stuff (C, Java, Python) to fun things (dlang, erlang, go). Platforms: I've played with Java's ANT build system, Google Appengine & AppsScript, Docker, LAMP & MEAN setups, built a CUPS server b/c windows machines can't print worth a damn, worked with key management (Ever used client SSL certs? They're amazing & ought to replace passwords), and (not quite so IT related) played with 3D printers to the point where I can design & build a moderately intricate dragon.
The IT job I'd like is basically what I have, but with a larger paycheck. I do sysadmin stuff at a small company that saw me at a programming competition in highschool. I work with ESXi VMs & I guess the biggest thing I've learned there is RAID setups.

**Comment by Jeffrey-P-McAteer** [comment]
Oh and of course I've done web stuff - html, js, css, php. Kinda important for OpenEMR dev. work.

**Comment by Jeffrey-P-McAteer** [comment]
Shell scripting too - almost too much. I recently turned a windows VM into a physical machine, which required turning the .vmdk file into a raw .img, then dd-ing the .img to a real disk. I didn't want to duplicate ~30GB though, so I made a FIFO file & set the virtualbox -> .img process to write to it & dd to read from it. Net result is no data ever touched my disk, it all streamed directly into the hdd.

**Comment by Jeffrey-P-McAteer** [comment]
IT skills I'd like to learn:
Haskell, more devopsian practices, I'd like to write my own FUSE-based filesystem.
Not a whole lot that could be integrated w/openemr, but I feel like that's the best.
I'm lacking in teamworking & time management skills more than anything, and while I may not get as excited about learning them they are much more important.

**Comment by MatthewVita** [comment]
@danielehrlich: Thanks for getting this started.

@Jeffrey-P-McAteer: You seem to have a great grip on all things Docker and general DevOps. Would you be interested in diving into Kubernetes for the version 2 of this project?

Just for reference, v1.0.0-ALPHA is just a bunch of manual steps to get OpenEMR in AWS without true HIPAA compliance. v1.0.0 proper will be HIPAA compliant and we may even use cloud formation to script everything out.

v2.0.0 is more interesting (to me anyways) as it will be all OpenEMR in Kubernetes (which makes the cloud provider irrelevant). One may even choose to run a "local cloud" if they are in a developing country with bad internet or just want to save money.

The Kubernetes tutorial is very good (I went through it the other day). A good start would be to add a ticket in our tracker for you to study Kubernetes. Another good ticket would be to expand on my Redis Docker work here: https://github.com/GoTeamEpsilon/OpenEMR-AWS-Guide/blob/master/Other/MRV_working_guide.MD#elasticache-memcached-sessions (would be great to be horizontally scaled in a Kubernetes service with stateful sets). There is a lot of stuff to think about here so let me know and I'll walk your through it so you know a) what the problem is and b) how to approach the problem with Kubernetes.

**Comment by MatthewVita** [comment]
Also, for career stuff, my advice is to consistently work on your goals each day (write down your goals and include them in your schedule). I've been doing this for years and it's played out well, idk.

Looking forward to our coding session tomorrow! EDIT: Oops, I mean Wednesday!!!

**Comment by MatthewVita** [comment]
@danielehrlich 

> another OS project we can start in parallel that matches to Jeffrey's skills

Hmm. I'm kind of out of ideas. We will eventually finish this project and jump on something else. I also don't see any reason why we couldn't start another project in tandem. Go for it!

**Comment by ghost** [comment]
1_. For Time Management I would say...first make sure you don't have ADHD. Hah Matt knows I'm big on this (let me get on my high horse here). I assume you have insurance since you're under 26 and can be with your parents so there are plenty of treatment options (there is 30% chance statistically since you're a male in the US).

2a. I would look into Terraform too. I think this has the potential to be the future. It's orchestration instead of configuration management (Chef, Ansible). You know the CloudFormation script we're writing? We'll it's like that but it works every where. @MatthewVita we haven't discussed this but maybe this can be v3.0 or maybe it's a dead end but @Jeffrey-P-McAteer I would check it out.

2b. A good "deliverable" would be writing a Terraform script that spins up say a server with specific security groups and installs a few programs with AWS. Would be cool for me to see that vs what our CF script looks like.

**Comment by MatthewVita** [comment]
@danielehrlich two quick questions:

1) Can Terraform run as a "local cloud"?
2) Does Terraform work with containers / come with a scheduler to orchestrate containers?

**Comment by ghost** [comment]
1. Not sure. It says it has a "Bare Metal" component, but not sure...the only reference I found online is Oracle's Bare Metal Cloud lol
2. Yes. I don't know about the scheduler though: https://www.terraform.io/docs/providers/kubernetes/index.html

**Comment by Jeffrey-P-McAteer** [comment]
I can't see terraform being capable of provisioning bare metal resources - you'd need some sort of management module, which are amazing but not all servers have them & APIs would be a mess.
I'm going to play with it & see how easily I can provision both AWS and google cloud resources.

**Comment by Jeffrey-P-McAteer** [comment]
That's not denying a "local cloud" though - no reason why you can't throw ESXi on a few machines & setup vCenter.

**Comment by MatthewVita** [comment]
> I'm going to play with it & see

Yeah, I was gonna say this is the best approach.

My money on K8S with Docker, but I need to leave the echo chamber for a bit here :)

**Comment by MatthewVita** [comment]
This work is now covered in our new issue/task tracker: https://app.asana.com/0/346554951567668/346554951567668
